cps/kni.c

```diff
@@ -850,6 +850,273 @@ rd_send_err(const struct nlmsghdr *req, int err)
+	int index;
+
+	int ret = rte_lpm6_iterator_state_init(ltbl->lpm6, 0, 0, &state6);
+	if (ret < 0) {
+		CPS_LOG(ERR, "Failed to initialize the IPv6 LPM rule iterator state in %s\n",
+			__func__);
+		return ret;
+	}
+
+	index = rte_lpm6_rule_iterate(&state6, &re6);
+	while (index >= 0) {
+		const struct in6_addr neighbor_gw = { 0 };
+		struct gk_fib *fib = &ltbl->fib_tbl6[re6.next_hop];
+		struct ipaddr *gw_addr;
+		struct nlmsghdr *reply;
+
+		if (fib->action != GK_FWD_GATEWAY_FRONT_NET &&
+				fib->action != GK_FWD_GATEWAY_BACK_NET &&
+				fib->action != GK_FWD_GRANTOR) {
+			index = rte_lpm6_rule_iterate(&state6, &re6);
+			continue;
+		}
+
+		reply = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+
+		rd_fill_getroute_reply(cps_conf, reply, fib,
+			family, req->nlmsg_seq, state6.depth, &gw_addr);
+
+		/* Add address. */
+		mnl_attr_put(reply, RTA_DST,
+			sizeof(struct in6_addr), re6.ip);
+
+		/*
+		 * If gateway is NULL, then the entry is for a
+		 * neighbor and the gateway should be ::.
+		 */
+		mnl_attr_put(reply, RTA_GATEWAY, sizeof(struct in6_addr),
+			gw_addr == NULL ? &neighbor_gw : &gw_addr->ip.v6);
+
+		if (!mnl_nlmsg_batch_next(batch)) {
+			ret = rd_send_batch(cps_conf, batch, "IPv6",
+				req->nlmsg_seq, req->nlmsg_pid, false);
+			if (ret < 0)
+				return ret;
+			mnl_nlmsg_batch_reset(batch);
+		}
+
+		index = rte_lpm6_rule_iterate(&state6, &re6);
+	}
+
+	return 0;
+}
+
+static int
+rd_getroute(const struct nlmsghdr *req, int *err)
+{
+	/*
+	 * Buffer length set according to libmnl documentation:
+	 * the buffer that you have to use to store the batch must be
+	 * double of MNL_SOCKET_BUFFER_SIZE to ensure that the last
+	 * message (message N+1) that did not fit into the batch is
+	 * written inside valid memory boundaries.
+	 */
+	char buf[2 * MNL_SOCKET_BUFFER_SIZE];
+	struct mnl_nlmsg_batch *batch;
+	struct cps_config *cps_conf = get_cps_conf();
+	struct gk_lpm *ltbl = &cps_conf->gk->lpm_tbl;
+	int family;
+
+	if (mnl_nlmsg_get_payload_len(req) < sizeof(struct rtgenmsg)) {
+		CPS_LOG(ERR, "Not enough room in CPS GETROUTE message from routing daemon in %s\n",
+			__func__);
+		*err = -EINVAL;
+		goto out;
+	}
+
+	family = ((struct rtgenmsg *)mnl_nlmsg_get_payload(req))->rtgen_family;
+
+	/* We don't support AF_UNSPEC to dump both tables at once. */
+	if (family != AF_INET && family != AF_INET6) {
+		CPS_LOG(ERR, "Unsupported address family type (%d) in %s\n",
+			family, __func__);
+		*err = -EAFNOSUPPORT;
+		goto out;
+	}
+
+	batch = mnl_nlmsg_batch_start(buf, MNL_SOCKET_BUFFER_SIZE);
+	if (batch == NULL) {
+		CPS_LOG(ERR, "Failed to allocate a batch for a GETROUTE reply\n");
+		*err = -ENOMEM;
+		goto out;
+	}
+
+	rte_spinlock_lock_tm(&ltbl->lock);
+	if (family == AF_INET) {
+		*err = rd_getroute_ipv4_locked(cps_conf, ltbl,
+			batch, req, family);
+	} else {
+		*err = rd_getroute_ipv6_locked(cps_conf, ltbl,
+			batch, req, family);
+	}
+	rte_spinlock_unlock_tm(&ltbl->lock);
+	if (*err < 0)
+		goto out;
```

Comment:
If we get an error in the middle of sending the LPM table, shouldn't we send the DONE message anyway? Isn't there a way to notify that there was an error in the middle of the listing?

Reply:
Based on the kernel code, it seems that if there is an error in the middle of sending the routes that whatever routes that are found are sent and no error message is present. I don't think there's a way to successfully report some entries and also indicate an error.

Reply:
Ok. We are following the kernel.

Reply:
On second look, it seems that you actually can notify the requester of both valid entries and an error because all of the messages sent back to the requester will be either of type `RTM_NEWROUTE`, `NLMSG_DONE`, or `NLMSG_ERROR` and they will match the sequence number of the request.

In the kernel it seems that all routes are sent back in one packet, followed by a separate error packet if one occurred. Luckily, this matches the structure we have now (send all updates, then call `rd_send_err()` in `rd_cb()` after that has completed.

So the way it is now, we send as many batches as needed, and if it fails we send an `NLMSG_ERROR` to let the requester know no more routes are available. Otherwise, we terminate the route list with an `NLMSG_DONE` message.

Reply:
The daemon stops listening for replies after reading with an `NLMSG_DONE` or `NLMSG_ERROR`.